README.md
<p align="center">
  <img alt="OS.js Logo" src="https://raw.githubusercontent.com/os-js/gfx/master/logo-big.png" />
</p>

[OS.js](https://www.os-js.org/) is an [open-source](https://raw.githubusercontent.com/os-js/OS.js/master/LICENSE) web desktop platform with a window manager, application APIs, GUI toolkit, filesystem abstractions and much more.

[![Support](https://img.shields.io/badge/patreon-support-orange.svg)](https://www.patreon.com/user?u=2978551&ty=h&u=2978551)
[![Support](https://img.shields.io/badge/opencollective-donate-red.svg)](https://opencollective.com/osjs)
[![Donate](https://img.shields.io/badge/liberapay-donate-yellowgreen.svg)](https://liberapay.com/os-js/)
[![Donate](https://img.shields.io/badge/paypal-donate-yellow.svg)](https://paypal.me/andersevenrud)

# OS.js Widgets module

This is the Widgets module for OS.js

* https://manual.os-js.org/v3/install/
* https://manual.os-js.org/v3/guide/provider/
* https://manual.os-js.org/v3/tutorial/widget/

## Installation

First, install the module:

```bash
npm install @osjs/widgets
```

In your client bootstrap (`src/client/index.js`):

```javascript
import {WidgetServiceProvider} from '@osjs/widgets';

osjs.register(WidgetServiceProvider);
```

And in your stylesheet (`src/client/index.scss`):

```css
@import "~@osjs/widgets/dist/main.css";
```

To set up a default set of widgets in the user settings, modify your client configuration file (`src/client/config.js`):

```javascript
{
  desktop: {
    settings: {
      widgets: [{
        name: 'digitalclock'
      }]
    }
  }
}
```

A contextmenu entry on the desktop is automatically added so users can add these themselves.

## Contribution

* **Sponsor on [Github](https://github.com/sponsors/andersevenrud)**
* **Become a [Patreon](https://www.patreon.com/user?u=2978551&ty=h&u=2978551)**
* **Support on [Open Collective](https://opencollective.com/osjs)**
* [Contribution Guide](https://github.com/os-js/OS.js/blob/master/CONTRIBUTING.md)

## Documentation

See the [Official Manuals](https://manual.os-js.org/) for articles, tutorials and guides.

## Links

* [Official Chat](https://gitter.im/os-js/OS.js)
* [Community Forums and Announcements](https://community.os-js.org/)
* [Homepage](https://os-js.org/)
* [Twitter](https://twitter.com/osjsorg) ([author](https://twitter.com/andersevenrud))
* [Facebook](https://www.facebook.com/os.js.org)
* [Docker Hub](https://hub.docker.com/u/osjs/)


CHANGELOG.md
# Changelog for osjs-widgets

## 3.1.0 - 2020-12-17

* Added automatic position clamp on init/resize (#15) (#16)

## 3.0.12 - 2020-11-26

* Updated dependencies

## 3.0.11 - 2020-10-13

* Added persian (fa_FA) translations (#13)

## 3.0.10 - 2020-09-22

* Minor cleanups

## 3.0.9 - 2020-09-22

* Added support for widget context menu titles (#11)
* Updated locales (#11)

## 3.0.8 - 2020-09-21

* Updated README
* Added locales (#11)

## 3.0.7 - 2020-07-31

* Updated default z-index

## 3.0.6 - 2020-07-17

* Changed root container usage
* Updated exports

## 3.0.5 - 2020-07-15

* Updated dependencies

## 3.0.3 - 2020-02-17

* Now using @osjs/dev-meta

## 3.0.2 - 2020-01-29

* Stop rendering when widget is destroyed (#3)
* Updated webpack configuration
* Updated dependencies
* Now using @osjs/stylelint-config
* Updated dotfile usage
* Updated copyright notices in preambles
* Updated package.json scripts
* Updated README
* Updated eslintrc
* Updated copyright(s)

## 3.0.1 - 2018-12-05

* Updated dependencies
* Fixed an invalid method call in digitalclock (#2)

## 3.0.0-alpha.4 - 2018-09-29

* Updated dependencies

## 3.0.0-alpha.3 - 2018-07-24

* Prebuild npm package

## 3.0.0-alpha.2 - 2018-07-18

* Added stylelintrc
* Added travis-ci badge to README
* Lint pass
* Added initial travis-ci config

## 3.0.0-alpha.0 - 2018-07-16

* Initial Release


index.js
/**
 * OS.js - JavaScript Cloud/Web Desktop Platform
 *
 * Copyright (c) 2011-2020, Anders Evenrud <andersevenrud@gmail.com>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this
 *    list of conditions and the following disclaimer
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
 * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 * @author  Anders Evenrud <andersevenrud@gmail.com>
 * @licence Simplified BSD License
 */

import './index.scss';
import WidgetServiceProvider from './src/provider';
import Widget from './src/widget';
import DigitalClockWidget from './src/items/digitalclock';

export {
  WidgetServiceProvider,
  Widget,
  DigitalClockWidget
};


src/locales.js
/*
 * OS.js - JavaScript Cloud/Web Desktop Platform
 *
 * Copyright (c) 2011-2020, Anders Evenrud <andersevenrud@gmail.com>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this
 *    list of conditions and the following disclaimer
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
 * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 * @author  Anders Evenrud <andersevenrud@gmail.com>
 * @licence Simplified BSD License
 */


export const en_EN = {
  LBL_ADD: 'Add Widget',
  LBL_REMOVE: 'Remove Widget',
  WIDGET_DIGITALCLOCK: 'Digital Clock'
};

export const nb_NO = {
  LBL_ADD: 'Legg til Widget',
  LBL_REMOVE: 'Fjern Widget',
  WIDGET_DIGITALCLOCK: 'Digitalklokke'
};


src/provider.js
/*
 * OS.js - JavaScript Cloud/Web Desktop Platform
 *
 * Copyright (c) 2011-2020, Anders Evenrud <andersevenrud@gmail.com>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this
 *    list of conditions and the following disclaimer
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
 * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 * @author  Anders Evenrud <andersevenrud@gmail.com>
 * @licence Simplified BSD License
 */

import DigitalClock from './items/digitalclock';
import * as translations from './locales';

/**
 * Widget Service Provider
 *
 * @desc Provides methods to handle widgets on a desktop in OS.js
 */
export default class WidgetServiceProvider {

  constructor(core, args = {}) {
    this.core = core;
    this.widgets = [];
    this.inited = false;
    this.registry = {digitalclock: DigitalClock, ...args.registry || {}};
  }

  destroy() {
    this.widgets.forEach(({widget}) => widget.destroy());
    this.widgets = [];
  }

  async init() {
    const iface = {
      register: (name, classRef) => {
        if (this.registry[name]) {
          console.warn('Overwriting previously registered widget item', name);
        }

        this.registry[name] = classRef;
      },

      removeAll: () => {
        this.widgets.forEach(({widget}) => widget.destroy());
        this.widgets = [];
      },

      remove: (widget) => {
        const index = typeof widget === 'number'
          ? widget
          : this.widgets.findIndex(w => w.widget === widget);

        if (index >= 0) {
          this.widgets[index].widget.destroy();
          this.widgets.splice(index, 1);
        }
      },

      create: (item) => {
        const ClassRef = this.registry[item.name];
        const widget = new ClassRef(this.core, item.options);
        this.widgets.push({name: item.name, widget});

        if (this.inited) {
          widget.init();
        }

        return widget;
      },

      get: (name) => this.registry[name],

      list: () => Object.keys(this.registry),

      save: () => {
        const settings = this.core.make('osjs/settings');
        const widgets = this.widgets.map(({name, widget}) => ({
          name,
          options: widget.options
        }));

        return Promise.resolve(settings.set('osjs/desktop', 'widgets', widgets))
          .then(() => settings.save());
      }
    };

    this.core.singleton('osjs/widgets', () => iface);

    this.core.on('osjs/desktop:transform', () => {
      this.widgets.forEach(({widget}) => widget.updatePosition(true));
    });
  }

  start() {
    this.inited = true;
    this.widgets.forEach(({widget}) => widget.init());

    const desktop = this.core.make('osjs/desktop');
    const __ = this.core
      .make('osjs/locale')
      .translatable(translations);

    let resizeDebounce;
    window.addEventListener('resize', () => {
      clearTimeout(resizeDebounce);
      resizeDebounce = setTimeout(() => this._clampWidgets(), 200);
    });

    if (typeof desktop.addContextMenuEntries === 'function') {
      desktop.addContextMenuEntries(() => {
        const widgets = this.core.make('osjs/widgets');

        return [{
          label: __('LBL_ADD'),
          items: widgets.list().map(t => {
            const classRef = this.registry[t];
            const metadata = classRef.metadata(this.core);

            return {
              label: metadata.title ? __(metadata.title) : t,
              onclick: () => {
                widgets.create({name: t});
                widgets.save();
              }
            };
          })
        }];
      });
    }

    this._clampWidgets();
  }

  _clampWidgets(resize) {
    if (resize && !this.core.config('windows.clampToViewport')) {
      return;
    }

    this.widgets.forEach(w => w.widget.clampToViewport());
  }
}


webpack.config.js
const path = require('path');
const MiniCssExtractPlugin = require('mini-css-extract-plugin');
const OptimizeCSSAssetsPlugin = require('optimize-css-assets-webpack-plugin');

const mode = process.env.NODE_ENV || 'development';
const minimize = mode === 'production';
const plugins = [];

if (mode === 'production') {
  plugins.push(new OptimizeCSSAssetsPlugin({
    cssProcessorOptions: {
      discardComments: true
    },
  }));
}

module.exports = {
  mode,
  devtool: 'source-map',
  entry: [
    path.resolve(__dirname, 'index.js'),
  ],
  output: {
    library: 'osjsWidgets',
    libraryTarget: 'umd',
    umdNamedDefine: true,
    sourceMapFilename: '[file].map',
    filename: '[name].js'
  },
  optimization: {
    minimize,
  },
  plugins: [
    new MiniCssExtractPlugin({
      filename: '[name].css',
      chunkFilename: '[id].css'
    }),
    ...plugins
  ],
  module: {
    rules: [
      {
        test: /\.(sa|sc|c)ss$/,
        exclude: /(node_modules|bower_components)/,
        use: [
          MiniCssExtractPlugin.loader,
          {
            loader: 'css-loader',
            options: {
              sourceMap: true
            }
          },
          {
            loader: 'sass-loader',
            options: {
              sourceMap: true
            }
          }
        ]
      },
      {
        test: /\.js$/,
        exclude: /(node_modules|bower_components)/,
        use: {
          loader: 'babel-loader'
        }
      }
    ]
  }
};


src/items/digitalclock.js
/*
 * OS.js - JavaScript Cloud/Web Desktop Platform
 *
 * Copyright (c) 2011-2020, Anders Evenrud <andersevenrud@gmail.com>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this
 *    list of conditions and the following disclaimer
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
 * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 * @author  Anders Evenrud <andersevenrud@gmail.com>
 * @licence Simplified BSD License
 */

import Widget from '../widget';

const getFont = (fontFamily, size) => String(size) + 'px ' + fontFamily;

const getTime = now => [
  now.getHours(),
  now.getMinutes(),
  now.getSeconds()
]
  .map(int => String(int).padStart(2, '0'))
  .join(':');

const getFontSize = (fontFamily, initialSize, maxWidth, context) => {
  const txt = '99:99:99';

  let size = initialSize;
  while (size > 0) {
    context.font = getFont(fontFamily, size);

    const measuredWidth = context.measureText(txt).width;
    if (measuredWidth < maxWidth) {
      break;
    }

    size--;
  }

  return size;
};

export default class DigitalClockWidget extends Widget {

  constructor(core, options) {
    super(core, options, {
      dimension: {
        width: 300,
        height: 50
      }
    }, {
      fontFamily: 'Monospace',
      fontColor: '#ffffff'
    });

    this.$tmpCanvas = document.createElement('canvas');
    this.tmpContext = this.$tmpCanvas.getContext('2d');
  }

  compute() {
    const {fontFamily, fontColor} = this.options;
    const {width, height} = this.$canvas;
    const {$tmpCanvas, tmpContext} = this;
    const size = getFontSize(fontFamily, height, width, tmpContext);

    $tmpCanvas.width = width;
    $tmpCanvas.height = size;

    tmpContext.font = getFont(fontFamily, size);
    tmpContext.fillStyle = fontColor;
    tmpContext.textAlign = 'center';
    tmpContext.textBaseline = 'middle';
  }

  onResize() {
    this.compute();
  }

  render({context, width, height}) {
    const {$tmpCanvas, tmpContext} = this;
    const tmpWidth = $tmpCanvas.width;
    const tmpHeight = $tmpCanvas.height;
    const x = (width / 2) - (tmpWidth / 2);
    const y = (height / 2) - (tmpHeight / 2);
    const text = getTime(new Date());

    tmpContext.clearRect(0, 0, tmpWidth, tmpHeight);
    tmpContext.fillText(text, tmpWidth / 2, tmpHeight / 2);
    context.clearRect(0, 0, width, height);
    context.drawImage($tmpCanvas, x, y, tmpWidth, tmpHeight);
  }

  getContextMenu() {
    return [{
      label: 'Set Color',
      onclick: () => this.createColorDialog()
    }, {
      label: 'Set Font',
      onclick: () => this.createFontDialog()
    }];
  }

  createFontDialog() {
    this.core.make('osjs/dialog', 'font', {
      name: this.options.fontFamily,
      controls: ['name']
    }, (btn, value) => {
      if (btn === 'ok') {
        this.options.fontFamily = value.name;
        this.compute();
        this.saveSettings();
      }
    });
  }

  createColorDialog() {
    this.core.make('osjs/dialog', 'color', {
      color: this.options.fontColor
    }, (btn, value) => {
      if (btn === 'ok') {
        this.options.fontColor = value.hex;
        this.compute();
        this.saveSettings();
      }
    });
  }

  static metadata() {
    return {
      ...super.metadata(),
      title: 'WIDGET_DIGITALCLOCK'
    }
  }
}


src/widget.js
/*
 * OS.js - JavaScript Cloud/Web Desktop Platform
 *
 * Copyright (c) 2011-2020, Anders Evenrud <andersevenrud@gmail.com>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this
 *    list of conditions and the following disclaimer
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
 * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 * @author  Anders Evenrud <andersevenrud@gmail.com>
 * @licence Simplified BSD License
 */

import merge from 'deepmerge';
import * as translations from './locales';

const MIN_WIDTH = 200;
const MIN_HEIGHT = 200;
const MAX_WIDTH = 800;
const MAX_HEIGHT = 800;
const DEFAULT_MARGIN = 15;
const EMIT_TIMEOUT = 44;
const isNull = val => typeof val === 'undefined' || val === null;

const getPosition = (core, position, nullValue = null) => {
  const rect = core.destroyed
    ? {}
    : core.make('osjs/desktop').getRect();

  let {top, left, right, bottom} = position;

  if (isNull(left) && isNull(right)) {
    left = rect.left + DEFAULT_MARGIN;
  } else if (!isNull(left)) {
    right = nullValue;
  } else {
    left = nullValue;
  }

  if (isNull(top) && isNull(bottom)) {
    top = rect.top + DEFAULT_MARGIN;
  } else if (!isNull(top)) {
    bottom = nullValue;
  } else {
    top = nullValue;
  }

  return {top, left, right, bottom};
};

const animator = (fps, stopFn) => {
  const interval = 1000 / fps;
  let last = performance.now();

  const animate = (cb) => {
    const now = performance.now();
    const elapsed = now - last;

    if (elapsed > interval) {
      last = now - (elapsed % interval);

      cb();
    }

    if (!stopFn()) {
      requestAnimationFrame(() => animate(cb));
    }
  };

  return animate;
};

const onmousedown = (ev, $root, widget) => {
  let debounce;
  const startX = ev.clientX;
  const startY = ev.clientY;
  const startPosition = {
    left: widget.$element.offsetLeft,
    top: widget.$element.offsetTop
  };
  const startDimension = {...widget.options.dimension};
  const resize = ev.target.classList.contains('osjs-widget-resize');
  const {minDimension, maxDimension} = widget.attributes;

  const mousemove = ev => {
    const diffX = ev.clientX - startX;
    const diffY = ev.clientY - startY;

    // TODO: Aspect Ratio

    clearTimeout(debounce);

    if (resize) {
      let newWidth = startDimension.width + diffX;
      let newHeight = startDimension.height + diffY;
      newWidth = Math.min(maxDimension.width, Math.max(minDimension.width, newWidth));
      newHeight = Math.min(maxDimension.height, Math.max(minDimension.height, newHeight));

      widget.options.dimension.width = newWidth;
      widget.options.dimension.height = newHeight;
      widget.updateDimension();
      debounce = setTimeout(() => widget.onResize(), EMIT_TIMEOUT);
    } else {
      widget.options.position.top = startPosition.top + diffY;
      widget.options.position.left = startPosition.left + diffX;
      widget.updatePosition();
      debounce = setTimeout(() => widget.onMove(), EMIT_TIMEOUT);
    }
  };

  const mouseup = ev => {
    window.removeEventListener('mousemove', mousemove);
    window.removeEventListener('mouseup', mouseup);

    widget.$element.classList.remove('active');
    $root.setAttribute('data-window-action', String(false));

    widget.saveSettings();
  };

  window.addEventListener('mousemove', mousemove);
  window.addEventListener('mouseup', mouseup);

  widget.$element.classList.add('active');
  $root.setAttribute('data-window-action', String(true));
};

export const clampPosition = (rect, {dimension, position}) => {
  const maxLeft = rect.width - dimension.width;
  const maxTop = rect.height - dimension.height;

  return {
    left: Math.max(0, Math.min(maxLeft, position.left)),
    top: Math.max(0, Math.max(rect.top, Math.min(maxTop, position.top)))
  };
};

export default class Widget {

  /**
   * @param {Object} options The options from the provider create function (usually settings storage result)
   * @param {Object} [attrs] Widget attributes
   * @param {Object} [settings] A set of defaults for the settings storage
   */
  constructor(core, options, attrs = {}, settings = {}) {
    this.core = core;
    this.dialog = null;
    this.$element = document.createElement('div');
    this.$canvas = document.createElement('canvas');
    this.context = this.$canvas.getContext('2d');
    this.destroyed = false;
    this.saveDebounce = null;
    this.attributes = {
      aspect: false,
      canvas: true,
      fps: 1,
      position: {
        top: null,
        left: null,
        right: null,
        bottom: null,
      },
      dimension: {
        width: MIN_WIDTH,
        height: MIN_HEIGHT,
      },
      minDimension: null,
      maxDimension: {
        width: MAX_WIDTH,
        height: MAX_HEIGHT
      }, ...attrs};

    if (this.attributes.minDimension === null) {
      this.attributes.minDimension = {
        width: MIN_WIDTH,
        height: MIN_HEIGHT,
        ...this.attributes.dimension
      };
    }

    if (this.attributes.aspect === true) {
      const {width, height} = this.attributes.dimension;
      const {maxDimension} = this.attributes;
      const aspect = width / height;

      this.attributes.aspect = aspect;
      this.attributes.minDimension.height = width / aspect;
      this.attributes.maxDimension.height = maxDimension.width * aspect;
    }

    this.options = {
      position: {...this.attributes.position},
      dimension: {...this.attributes.dimension},
      ...settings,
      ...options
    };
  }

  destroy() {
    if (this.destroyed) {
      return;
    }

    this.destroyed = true;
    this.onDestroy();

    this.saveDebounce = clearTimeout(this.saveDebounce);

    if (this.dialog) {
      this.dialog.destroy();
    }

    if (this.$element) {
      if (this.$element.parentNode) {
        this.$element.remove();
      }
    }

    this.$canvas = null;
    this.$element = null;
    this.context = null;
    this.dialog = null;
    this.options = {};
    this.attributes = {};
  }

  /**
   * @param {Object} options
   * @param {number} options.width
   * @param {number} options.height
   * @param {HTMLCanvasElement} options.canvas
   * @param {CanvasRenderingContext2D} options.context
   */
  render(viewport) {
  }

  start() {
    const render = () => this.render({
      width: this.options.dimension.width,
      height: this.options.dimension.height,
      canvas: this.$canvas,
      context: this.context
    });

    this.updateDimension();
    this.updatePosition();
    this.onResize();
    this.onMove();

    render();

    if (this.attributes.canvas) {
      animator(this.attributes.fps, () => this.destroyed)(() => render());
    }

    setTimeout(() => this.clampToViewport(), 100);
  }

  init() {
    const $root = this.core.$contents;
    const resizer = document.createElement('div');
    resizer.classList.add('osjs-widget-resize');

    this.$element.appendChild(resizer);
    this.$element.addEventListener('mousedown', ev => onmousedown(ev, $root, this));
    this.$element.addEventListener('contextmenu', ev => this.onContextMenu(ev));
    this.$element.classList.add('osjs-widget');
    $root.appendChild(this.$element);

    if (this.attributes.canvas) {
      this.$element.appendChild(this.$canvas);
    }

    this.start();
  }

  updateDimension() {
    if (this.destroyed) {
      return;
    }

    const {width, height} = this.options.dimension;
    this.$element.style.width = String(width) + 'px';
    this.$element.style.height = String(height) + 'px';
    this.$canvas.width = width;
    this.$canvas.height = height;
  }

  updatePosition() {
    if (this.destroyed) {
      return;
    }

    const {left, right, top, bottom} = getPosition(this.core, this.options.position, 'auto');
    const getValue = val => typeof val === 'string' ? val : `${val}px`;

    this.$element.style.left = getValue(left);
    this.$element.style.right = getValue(right);
    this.$element.style.top = getValue(top);
    this.$element.style.bottom = getValue(bottom);
  }

  saveSettings() {
    this.saveDebounce = clearTimeout(this.saveDebounce);
    this.saveDebounce = setTimeout(() => this.core.make('osjs/widgets').save(), 100);
  }

  getContextMenu() {
    return [];
  }

  onDestroy() {
  }

  onResize() {
  }

  onMove() {
  }

  onContextMenu(ev) {
    const __ = this.core
      .make('osjs/locale')
      .translatable(translations);

    const menu = [
      ...this.getContextMenu(),
      {
        label: __('LBL_REMOVE'),
        onclick: () => {
          this.core.make('osjs/widgets')
            .remove(this);

          this.core.make('osjs/widgets')
            .save();
        }
      }
    ];

    this.core.make('osjs/contextmenu').show({
      position: ev,
      menu
    });
  }

  _createDialog(options, callbackRender, callbackValue) {
    if (this.dialog) {
      return false;
    }

    const callbackButton = (btn, options) => {
      if (btn === 'ok') {
        this.options = merge(this.options, options);
        this.saveSettings();
      }
    };

    const dialogOptions = {
      buttons: ['ok', 'cancel'],
      window: options || {}
    };

    this.dialog = this.core.make('osjs/dialogs')
      .create(dialogOptions, callbackValue, callbackButton)
      .render(callbackRender);

    this.dialog.win.on('destroy', () => (this.dialog = null));

    return this.dialog;
  }

  clampToViewport() {
    const {top, left} = this.options.position;
    const rect = this.core.make('osjs/desktop').getRect();
    const pos = clampPosition(rect, this.options);

    if (pos.left !== left || pos.top !== top) {
      this.options.position = {...pos};
      this.updatePosition();
    }
  }

  static metadata() {
    return {
      title: undefined
    };
  }
}
